apollo-federation/src/sources/connect/validation/selection/visitor.rs

```diff
@@ -0,0 +1,288 @@
+use crate::sources::connect::expand::visitors::FieldVisitor;
+use crate::sources::connect::expand::visitors::GroupVisitor;
+use crate::sources::connect::json_selection::LitExpr;
+use crate::sources::connect::json_selection::MethodArgs;
+use crate::sources::connect::json_selection::NamedSelection;
+use crate::sources::connect::json_selection::PathList;
+use crate::sources::connect::JSONSelection;
+use crate::sources::connect::SubSelection;
+
+/// A JSON Selection visitor
+pub(super) trait Visitor {
+    type Error;
+
+    /// Visit a JSON Selection part. The visitor is guaranteed that all children of this part will
+    /// be visited before [`Visitor::end_visit`] is called. A mutable reference is required since
+    /// the visitor will typically require modifying internal state.
+    fn visit(&mut self, part: &SelectionPart) -> Result<(), Self::Error>;
+
+    /// End visiting a JSON Selection part. This is called after all children of a part have been
+    /// visited. The default implementation does nothing.
+    fn end_visit(&mut self, _part: &SelectionPart) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+/// A part of a JSON Selection to be visited
+#[derive(Clone, Debug)]
+pub(super) enum SelectionPart<'schema> {
+    JSONSelection(&'schema JSONSelection),
+    LitExpr(&'schema LitExpr),
+    MethodArgs(&'schema MethodArgs),
+    NamedSelection(&'schema NamedSelection),
+    PathList(&'schema PathList),
+    SubSelection(&'schema SubSelection),
+}
```

Comment:
What qualifies a given AST structure as a part of the selection that should be visited, since this list is not exhaustive? I'd suggest adding more detail to the comment.

Reply:
Just that I needed it to locate arrow methods. This is really just the beginning of a proper AST visitor to allow validating arrow methods. I think @nicholascioli is working on designing something more extensive.